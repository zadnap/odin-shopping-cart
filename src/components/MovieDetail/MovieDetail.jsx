import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styles from './MovieDetail.module.scss';
import Button from '@/components/Button/Button';
import Modal from '@/components/Modal/Modal';
import VideoPlayer from '@/components/VideoPlayer/VideoPlayer';
import {
  faPlay,
  faHeart,
  faCartPlus,
  faStar,
} from '@fortawesome/free-solid-svg-icons';
import { useState } from 'react';

function MovieDetail({ movie }) {
  const {
    title,
    posterSrc,
    backdropSrc,
    certification,
    releaseDate,
    duration,
    overview,
    genres,
    rating,
    languages,
    directors,
    writers,
    trailerKey,
    rent,
  } = movie;
  const [isOpenModal, setIsOpenModal] = useState(false);

  return (
    <article
      className={styles.movieDetail}
      style={{
        backgroundImage: `url(${backdropSrc})`,
      }}
    >
      <img
        className={styles.poster}
        src={posterSrc}
        alt={`${title}'s poster`}
      />
      <div className={styles.info}>
        <div className={styles.generalInfo}>
          <div className={styles.header}>
            <h1 className={styles.title}>{title}</h1>
            <div className={styles.rating}>
              <FontAwesomeIcon icon={faStar} /> <span>{rating}</span>
            </div>
          </div>
          <p className={styles.movieMeta}>
            <span className={styles.certification}>{certification}</span>{' '}
            &middot; {releaseDate} &middot; {duration}
          </p>
        </div>
        <ul className={styles.detailList}>
          <li className={styles.item}>
            <h2 className={styles.title}>Overview</h2>
            <p className={styles.description}>{overview}</p>
          </li>
          <li className={styles.item}>
            <h2 className={styles.title}>Genres</h2>
            <p className={styles.description}>
              {genres.map((genre) => (
                <span key={genre.id} className={styles.tag}>
                  {genre.name}
                </span>
              ))}
            </p>
          </li>
          {directors.length > 0 && (
            <li className={styles.item}>
              <h2 className={styles.title}>Directors</h2>
              <p className={styles.description}>
                {directors.map((director) => (
                  <span key={director} className={styles.tag}>
                    {director}
                  </span>
                ))}
              </p>
            </li>
          )}
          {writers.length > 0 && (
            <li className={styles.item}>
              <h2 className={styles.title}>Story</h2>
              <p className={styles.description}>
                {writers.map((writer) => (
                  <span key={writer} className={styles.tag}>
                    {writer}
                  </span>
                ))}
              </p>
            </li>
          )}
          <li className={styles.item}>
            <h2 className={styles.title}>Language</h2>
            <p className={styles.description}>
              {languages.map((lang) => (
                <span key={lang.id} className={styles.tag}>
                  {lang.name}
                </span>
              ))}
            </p>
          </li>
        </ul>
        <div className={styles.buttonGroup}>
          <Button>
            <FontAwesomeIcon icon={faCartPlus} /> Rent: ${rent}
          </Button>
          <Button>
            <FontAwesomeIcon icon={faHeart} /> Add to Favourites
          </Button>
          <Button onClick={() => setIsOpenModal(true)}>
            <FontAwesomeIcon icon={faPlay} /> Play trailer
          </Button>
        </div>
      </div>
      {isOpenModal && (
        <Modal
          title={`${title}'s trailer`}
          onClose={() => setIsOpenModal(false)}
        >
          <VideoPlayer videoKey={trailerKey} movieTitle={title} />
        </Modal>
      )}
    </article>
  );
}

export default MovieDetail;
