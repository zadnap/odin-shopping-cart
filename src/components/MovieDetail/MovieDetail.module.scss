@use '@/styles/variables.scss' as *;

.movieDetail {
  position: relative;
  border-radius: $border-radius;
  padding: 15px;
  display: flex;
  gap: 15px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  z-index: 1;
  height: calc(100vh - $header-height);

  &::before {
    border-radius: $border-radius;
    position: absolute;
    content: '';
    background-color: $overlay-color;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    z-index: -1;
  }

  .poster {
    border-radius: $border-radius;
    object-fit: cover;
    height: 100%;
  }

  .info {
    display: flex;
    flex-direction: column;
    gap: 20px;
    overflow-y: auto;
    scrollbar-width: none;
    -ms-overflow-style: none;

    &::-webkit-scrollbar {
      display: none;
    }

    .generalInfo {
      .header {
        display: flex;
        align-items: flex-start;
        justify-content: space-between;

        .title {
          font-size: 2.5rem;
        }
        .rating {
          font-size: 1.5rem;
          padding: 4px 10px;
          color: yellow;
          border-radius: $border-radius;
          background-color: $overlay-color;
        }
      }

      .movieMeta {
        font-size: 1.5rem;
        font-weight: 300;

        .certification {
          padding: 1px 5px;
          border: 1px solid $subtext-color;
        }
      }
    }

    .detailList {
      display: flex;
      flex-direction: column;
      gap: 15px;

      .item {
        .title {
          font-size: 1.6rem;
          font-weight: 500;
        }
        .description {
          font-size: 1.5rem;
          display: flex;
          flex-wrap: wrap;
          gap: 5px;

          .tag {
            padding: 3px 8px;
            border-radius: $border-radius;
            background-color: $overlay-color;
          }
        }
      }
    }

    .buttonGroup {
      display: flex;
      justify-content: end;
      gap: 8px;
      margin-top: auto;
    }
  }
}
