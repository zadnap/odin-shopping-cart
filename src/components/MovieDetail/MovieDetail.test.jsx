import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import MovieDetail from './MovieDetail';

const mockProps = {
  title: 'Lilo & Stitch',
  posterSrc:
    'https://www.themoviedb.org/t/p/w600_and_h900_bestv2/7c5VBuCbjZOk7lSfj9sMpmDIaKX.jpg',
  backdropSrc:
    'https://image.tmdb.org/t/p/original/7Zx3wDG5bBtcfk8lcnCWDOLM4Y4.jpg',
  certification: 'PG',
  releaseDate: '05/23/2025',
  duration: '1h 48m',
  overview:
    'The wildly funny and touching story of a lonely Hawaiian girl and the fugitive alien who helps to mend her broken family.',
  genres: [
    { id: 1, name: 'Family' },
    { id: 2, name: 'Science' },
    { id: 3, name: 'Fiction' },
    { id: 4, name: 'Comedy' },
    { id: 5, name: 'Adventure' },
  ],
  rating: 7.1,
  language: 'English',
  directors: ['Dean Fleischer Camp'],
  writers: ['Chris Kekaniokalani Bright', 'Mike Van Waes'],
  rent: '$4.99',
};

describe('MovieDetail component', () => {
  it('should render to match snapshot', () => {
    const { container } = render(<MovieDetail {...mockProps} />);

    expect(container).toMatchSnapshot();
  });
});
