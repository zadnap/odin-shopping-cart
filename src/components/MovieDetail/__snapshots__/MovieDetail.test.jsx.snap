// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieDetail component > should render to match snapshot 1`] = `
<div>
  <article
    class="_movieDetail_d06782"
    style="background-image: url("https://image.tmdb.org/t/p/original/7Zx3wDG5bBtcfk8lcnCWDOLM4Y4.jpg");"
  >
    <img
      alt="Lilo & Stitch's poster"
      class="_poster_d06782"
      src="https://www.themoviedb.org/t/p/w600_and_h900_bestv2/7c5VBuCbjZOk7lSfj9sMpmDIaKX.jpg"
    />
    <div
      class="_info_d06782"
    >
      <div
        class="_generalInfo_d06782"
      >
        <div
          class="_header_d06782"
        >
          <h1
            class="_title_d06782"
          >
            Lilo & Stitch
          </h1>
          <div
            class="_rating_d06782"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-star "
              data-icon="star"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                fill="currentColor"
              />
            </svg>
             
            <span>
              7.1
            </span>
          </div>
        </div>
        <p
          class="_movieMeta_d06782"
        >
          <span
            class="_certification_d06782"
          >
            PG
          </span>
           
          · 
          05/23/2025
           · 
          1h 48m
        </p>
      </div>
      <ul
        class="_detailList_d06782"
      >
        <li
          class="_item_d06782"
        >
          <h2
            class="_title_d06782"
          >
            Overview
          </h2>
          <p
            class="_description_d06782"
          >
            The wildly funny and touching story of a lonely Hawaiian girl and the fugitive alien who helps to mend her broken family.
          </p>
        </li>
        <li
          class="_item_d06782"
        >
          <h2
            class="_title_d06782"
          >
            Genres
          </h2>
          <p
            class="_description_d06782"
          >
            <span
              class="_tag_d06782"
            >
              Family
            </span>
            <span
              class="_tag_d06782"
            >
              Science
            </span>
            <span
              class="_tag_d06782"
            >
              Fiction
            </span>
            <span
              class="_tag_d06782"
            >
              Comedy
            </span>
            <span
              class="_tag_d06782"
            >
              Adventure
            </span>
          </p>
        </li>
        <li
          class="_item_d06782"
        >
          <h2
            class="_title_d06782"
          >
            Directors
          </h2>
          <p
            class="_description_d06782"
          >
            <span
              class="_tag_d06782"
            >
              Dean Fleischer Camp
            </span>
          </p>
        </li>
        <li
          class="_item_d06782"
        >
          <h2
            class="_title_d06782"
          >
            Story
          </h2>
          <p
            class="_description_d06782"
          >
            <span
              class="_tag_d06782"
            >
              Chris Kekaniokalani Bright
            </span>
            <span
              class="_tag_d06782"
            >
              Mike Van Waes
            </span>
          </p>
        </li>
        <li
          class="_item_d06782"
        >
          <h2
            class="_title_d06782"
          >
            Language
          </h2>
          <p
            class="_description_d06782"
          >
            <span
              class="_tag_d06782"
            >
              English
            </span>
          </p>
        </li>
      </ul>
      <div
        class="_buttonGroup_d06782"
      >
        <button
          class="
        _button_ee73f7 
         
         
        
      "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cart-plus "
            data-icon="cart-plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 24C0 10.7 10.7 0 24 0L69.5 0c22 0 41.5 12.8 50.6 32l411 0c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3l-288.5 0 5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5L488 336c13.3 0 24 10.7 24 24s-10.7 24-24 24l-288.3 0c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5L24 48C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96zM252 160c0 11 9 20 20 20l44 0 0 44c0 11 9 20 20 20s20-9 20-20l0-44 44 0c11 0 20-9 20-20s-9-20-20-20l-44 0 0-44c0-11-9-20-20-20s-20 9-20 20l0 44-44 0c-11 0-20 9-20 20z"
              fill="currentColor"
            />
          </svg>
           Rent: $
          $4.99
        </button>
        <button
          class="
        _button_ee73f7 
         
         
        
      "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart "
            data-icon="heart"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
              fill="currentColor"
            />
          </svg>
           Add to Favourites
        </button>
        <button
          class="
        _button_ee73f7 
         
         
        
      "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80L0 432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"
              fill="currentColor"
            />
          </svg>
           Play trailer
        </button>
      </div>
    </div>
  </article>
</div>
`;
